google/cloud/storage/client.cc

```diff
@@ -184,6 +186,37 @@ void Client::DownloadFileImpl(internal::ReadObjectRangeRequest const& request,
   }
 }
 
+std::string Client::SignUrl(internal::SignUrlRequest const& request) {
+  auto base_credentials = raw_client()->client_options().credentials();
+  auto credentials = dynamic_cast<oauth2::ServiceAccountCredentials<>*>(
+      base_credentials.get());
+
+  if (credentials == nullptr) {
+    google::cloud::internal::RaiseRuntimeError(
+        R"""(The current credentials cannot be used to sign URLs.
+Please configure your google::cloud::storage::Client to use service account
+credentials, as described in:
+https://cloud.google.com/storage/docs/authentication
+)""");
+  }
+
+  auto result = credentials->SignString(request.StringToSign());
```

Comment:
This is missing support for ADC SignString using the IAM API to signBlob:
https://cloud.google.com/iam/reference/rest/v1/projects.serviceAccounts/signBlob

ADC signing would be used in Compute Engine and GAE flexible environment custom runtimes when a SA private key pem is not available.

Reply:
Created #1595 for that.